src/libraries/System.Private.CoreLib/src/System/ComAwareWeakReference.cs

```diff
@@ -0,0 +1,157 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+#if FEATURE_COMINTEROP || FEATURE_COMWRAPPERS
+
+using static System.WeakReferenceHandleTags;
+
+namespace System
+{
+    internal sealed partial class ComAwareWeakReference
+    {
+        private readonly nint _weakHandle;
+        private ComInfo? _comInfo;
+
+        internal sealed class ComInfo
+        {
+            private readonly IntPtr _pComWeakRef;
+            private readonly long _wrapperId;
+
+            internal object? ResolveTarget()
+            {
+                return ComWeakRefToObject(_pComWeakRef, _wrapperId);
+            }
+
+            // see: syncblk.h
+            private const int IS_HASHCODE_BIT_NUMBER = 26;
+            private const int BIT_SBLK_IS_HASHCODE = 1 << IS_HASHCODE_BIT_NUMBER;
+            private const int BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX = 0x08000000;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal static unsafe ComInfo? FromObject(object? target)
+            {
+                if (target == null)
+                    return null;
+
+                fixed (byte* pRawData = &target.GetRawData())
+                {
+                    // The header is 4 bytes before MT field on all architectures
+                    int header = *(int*)(pRawData - sizeof(IntPtr) - sizeof(int));
+                    // common case: target does not have a syncblock, so there is no interop info
+                    if ((header & (BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX | BIT_SBLK_IS_HASHCODE)) != BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX)
+                        return null;
+                }
+
+                return FromObjectSlow(target);
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            private static unsafe ComInfo? FromObjectSlow(object target)
+            {
+                IntPtr pComWeakRef = ObjectToComWeakRef(target, out long wrapperId);
+                if (pComWeakRef == 0)
+                    return null;
+
+                return new ComInfo(pComWeakRef, wrapperId);
+            }
+
+            private ComInfo(IntPtr pComWeakRef, long wrapperId)
+            {
+                Debug.Assert(pComWeakRef != IntPtr.Zero);
+                _pComWeakRef = pComWeakRef;
+                _wrapperId = wrapperId;
+            }
+
+            ~ComInfo()
+            {
+                Marshal.Release(_pComWeakRef);
```

Comment:
This assumes that the COM reference is free-threaded and that is ok to call Release on the finalizer thread.

Does the existing code make the same assumption?

Reply:
No, or at least it shouldn't. The COM `Release` on the finalizer goes through the context dispatch logic contained in `SyncBlock` cleanup. See below.

https://github.com/dotnet/runtime/blob/33cb335386d39da8e4e89e75449628be7e3b375a/src/coreclr/vm/syncblk.cpp#L542

Reply:
>  is ok to call Release on the finalizer thread.

I think there is some confusion here.
This is not the release for the interop info that lives in the sync block. That part did not change in this PR. 
This is for the `IWeakReference` pointer that we keep on a side in case we need to rehydrate the RCW. 

I do not think the current implementation makes any assumptions on which thread this Release is called. It runs as a part of WeakReference`s Finalize on whichever thread does the finalization.

Currently it can run on:
1. With Workstation GC: this Release runs on some random user thread that happens to do current GC.
2. With Server GC: this Release runs by multiple GC threads in parallel.  
If the Release may take locks, it is a bit of a concern as it will block/serialize the GC threads.
3. If WeakReference is inherited in another type, then eager finalization is not used and Finalize will run on Finalizer thread.

We can easily make `ComInfo` be eagerly finalized and it will match behavior in `#2` and `#3`. However I think there are no particular benefits from that. 

Also, since we are calling some external code that we do not own, calling it from inside the GC while EE is blocked feels like inviting some unnecessary risk. - What if this Release takes nontrivial time, what if it blocks, cannot tolerate concurrent calls, crashes or calls some managed code and reenters?
It feels like Finalizer thread can tolerate more abuse if Release misbehaves.

Reply:
> I do not think the current implementation makes any assumptions on which thread this Release is called.

Agreed. This does look fine.

> Also, since we are calling some external code that we do not own, calling it from inside the GC while EE is blocked feels like inviting some unnecessary risk. - What if this Release takes nontrivial time, what if it blocks, cannot tolerate concurrent calls or calls some managed code?

All fair questions. Running the CsWinRT test suite and working with that team seems appropriate to determine this and verify this change has not unexpected fall out. If you ping me offline, I can provide those contacts.

Reply:
I have run [CsWinRT](https://github.com/microsoft/CsWinRT) tests with this change and did not see any failures in 100 iterations over all tests under “UnitTests” node.